pkg/genericapiserver/server/serve.go

```diff
@@ -224,22 +228,187 @@ func getNamedCertificateMap(certs []namedTlsCert) (map[string]*tls.Certificate,
 	return byName, nil
 }
 
+// tcpConnectionChainListener is responsible to invoking any connection handling
+// code which is invoked after AcceptTCP() has been called but before the
+// connection is usable by the server for communicating with the client. We
+// assume if an error is returned from a chain handler that it will return a nil
+// conn and have already closed that conn.
+type tcpConnectionChainHandler interface {
+	init(*net.TCPConn) error
+	wrap(net.Conn) (net.Conn, error)
+}
+
+type tcpConnectionChainListener struct {
+	*net.TCPListener
+	chain []tcpConnectionChainHandler
+}
+
+func (l tcpConnectionChainListener) Accept() (net.Conn, error) {
+	tc, err := l.AcceptTCP()
+	if err != nil {
+		return nil, err
+	}
+	for _, handler := range l.chain {
+		err = handler.init(tc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var conn net.Conn = tc
+	for _, handler := range l.chain {
+		conn, err = handler.wrap(conn)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return tc, nil
+}
+
 // tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
 // connections. It's used by ListenAndServe and ListenAndServeTLS so
 // dead TCP connections (e.g. closing laptop mid-download) eventually
 // go away.
 //
 // Copied from Go 1.7.2 net/http/server.go
-type tcpKeepAliveListener struct {
-	*net.TCPListener
+type tcpKeepAliveHandler struct {
+}
+
+func (ln tcpKeepAliveHandler) init(conn *net.TCPConn) error {
+	conn.SetKeepAlive(true)
+	conn.SetKeepAlivePeriod(defaultKeepAlivePeriod)
+	return nil
+}
+
+func (ln tcpKeepAliveHandler) wrap(conn net.Conn) (net.Conn, error) {
+	return conn, nil
 }
 
-func (ln tcpKeepAliveListener) Accept() (net.Conn, error) {
-	tc, err := ln.AcceptTCP()
+// ipBasedLimitListener accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
```

Comment:
0.0.0.0 leads to 127.0.0.1 for loopback client as target IP. You have to bind to the IP of one of your interfaces, e.g. 192.168.2.42 if that is the eth0 IP. Then the apiserver will use that one to connect to itself. Then check what the source address is.